rust/tw_evm/src/abi/function.rs

```diff
@@ -0,0 +1,87 @@
+// Copyright © 2017-2023 Trust Wallet.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+use crate::abi::decode::decode_params;
+use crate::abi::param::Param;
+use crate::abi::param_token::NamedToken;
+use crate::abi::token::Token;
+use crate::abi::{AbiError, AbiErrorKind, AbiResult};
+use serde::Deserialize;
+use tw_memory::Data;
+
+#[derive(Clone, Debug, Default, Deserialize)]
+pub struct Function {
+    /// Function name.
+    pub name: String,
+    /// Function input.
+    pub inputs: Vec<Param>,
+    /// Function output.
+    pub outputs: Vec<Param>,
+    /// Whether the function reads or modifies blockchain state.
+    #[serde(rename = "stateMutability", default)]
+    pub state_mutability: ethabi::StateMutability,
+}
+
+impl Function {
+    /// Returns a signature that uniquely identifies this function including `inputs` **only**.
+    ///
+    /// Examples:
+    /// - `functionName()`
+    /// - `functionName()`
+    /// - `functionName(bool)`
+    /// - `functionName(uint256,bytes32)`
+    ///
+    /// The method implementation is inspired by
+    /// https://github.com/rust-ethereum/ethabi/blob/b1710adc18f5b771d2d2519c87248b1ba9430778/ethabi/src/function.rs#L88-L97
+    pub fn signature_with_inputs(&self) -> String {
```

Comment:
i guess `siganture_with_inputs` hints that this function generate signature only for function that have inputs which is not the case based on the implementation, there are functions with no `inputs` and would look like this `functionName()` (you already added that as an example in docs`

Reply:
Good note. Yes, I don't like the name of the function as well. The requirement not to put `outputs` to the function signature comes from the legacy implementation. But in general, we should be able to return the true function signature with inputs and outputs **if the specified**.
I'd refactor the function to include function `outputs` if only they are, so delegate the responsibility to specify or not to specify outputs to the function caller.

Reply:
Done